client/api/src/storage_changes_for_keys.rs

```diff
@@ -0,0 +1,609 @@
+use std::{
+	collections::{HashMap, HashSet, VecDeque},
+	pin::Pin,
+	sync::Arc,
+	task::Poll,
+};
+
+#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug)]
+#[repr(transparent)]
+struct SubscriberId(u64);
+
+#[derive(Debug)]
+enum NotificationKind {
+	/// The initial notification immediately after the subscriber was created..
+	Initial,
+	/// This new block immediately follows the block we've last seen.
+	Incremental,
+	/// This new block doesn't immediately follow the block we've last seen.
+	Reorganized,
+}
+
+/// An opaque notification about storage changes.
+#[derive(Debug)]
+pub struct StorageNotification<Hash> {
+	/// The hash of the block.
+	block_hash: Hash,
+
+	/// A list of storage changes.
+	///
+	/// This is unfiltered and contains *all* of the changes in the block.
+	all_changes: Arc<HashMap<StorageKey, Option<StorageData>>>,
+
+	/// The type of notification.
+	kind: NotificationKind,
+
+	/// A set of keys to which this subscriber is subscribed to.
+	subscribed_for: Arc<HashSet<StorageKey>>,
+}
+
+impl<Hash> StorageNotification<Hash> {
+	/// Extracts the storage change set from this notification and/or the storage.
+	pub fn get<Block, Backend, Storage>(self, client: &Storage) -> Option<StorageChangeSet<Hash>>
+	where
+		Storage: crate::StorageProvider<Block, Backend>,
+		Backend: crate::backend::Backend<Block>,
+		Block: BlockT<Hash = Hash>,
+		Hash: Clone,
+	{
+		let block = BlockId::Hash(self.block_hash.clone());
+		self.get_impl(|key| client.storage(&block, key).ok().flatten())
+	}
+
+	fn get_impl(
+		&self,
+		mut get_from_storage: impl FnMut(&StorageKey) -> Option<StorageData>,
+	) -> Option<StorageChangeSet<Hash>>
+	where
+		Hash: Clone,
+	{
+		let mut changes = Vec::new();
+		if matches!(self.kind, NotificationKind::Incremental) {
+			// The new block immediately follows the previous block
+			// so we don't have to fetch anything extra from the storage;
+			// we only need to grab the changes we're interested in.
+			changes = self
+				.all_changes
+				.iter()
+				.flat_map(|(key, value)| {
+					if self.subscribed_for.contains(&key) {
+						Some((key.clone(), value.clone()))
+					} else {
+						None
+					}
+				})
+				.collect();
+		} else {
+			// The new block doesn't immediately follow the previous block
+			// (or is the very first notification) so we have to fetch
+			// the values from the storage.
+			for key in &*self.subscribed_for {
+				if let Some(value) = self.all_changes.get(key) {
+					// One of the keys we care about was changed in this block,
+					// so no need to fetch it from the storage.
+					changes.push((key.clone(), value.clone()));
+				} else {
+					// The key we care about wasn't changed here, but it might
+					// have been changed in one of the previous blocks,
+					// so fetch it from the storage.
+					let value = get_from_storage(key);
+					if matches!(self.kind, NotificationKind::Initial) && value.is_none() {
+						// Don't send empty values for the initial notification.
```

Comment:
Is there any reason we are not sending `None` for initial Notifications? I see that it is not necessary, but feels like it would be more consistent to always send `None` values.

Reply:
No particular reason besides that it shouldn't be necessary.

I don't really have a strong opinion regarding this though so if the consensus is to send those I can change this.

Reply:
I also don't have a strong opinion on it, its a minor issue. It would just save us this extra condition and subscribers can build a perfectly valid state by just upserting all keys that come via RPC notification.